alf/algorithms/vae.py

```diff
@@ -179,3 +185,179 @@ def train_step(self, inputs, state=()):
+    the parameter gradient is
+
+    .. math::
+
+        \frac{\partial L}{\partial \hat{z}}\frac{\partial \hat{z}}{\partial \theta}
+        = \frac{\partial L}{\partial \hat{z}}\frac{\partial z_{prob}}{\partial \theta}
+
+    Alternatively, we provide the option of ST Gumbel Softmax gradient estimator.
+
+    ::
+
+        Jang et al., "CATEGORICAL REPARAMETERIZATION WITH GUMBEL-SOFTMAX", 2017.
+
+    Which applies the above ST trick to the Gumbel-softmax distribution that uses
+    the Gumbel trick to reparameterize the categorical sampling process. The paper
+    claims that ST Gumbel-softmax gradient estimator has a lower variance than the
+    plain ST estimator.
+    """
+
+    def __init__(self,
+                 z_spec: BoundedTensorSpec,
+                 input_tensor_spec: alf.NestedTensorSpec = None,
+                 z_network_cls: Callable = EncodingNetwork,
+                 prior_input_tensor_spec: alf.NestedTensorSpec = None,
+                 prior_z_network_cls: Callable = None,
+                 mode: str = "st",
+                 gumbel_temp_scheduler: Scheduler = ConstantScheduler(1.),
+                 beta: float = 1.,
+                 target_kld_per_categorical: float = None,
+                 beta_optimizer: torch.optim.Optimizer = None,
+                 name: str = "DiscreteVAE"):
+        """
+        Args:
+            z_spec: a tensor spec for the discrete posterior. Regardless of its
+                rank, it will be converted to rank-1, representing a vector of
+                discrete variables. The value bould of each variable must be
+                identical.
+            input_tensor_spec: the input spec.
+            z_network_cls: an encoding network to encode input data into a vector
+                of logits. If ``prior_z_network_cls`` is None, this network must
+                handle input with spec ``input_tensor_spec``. If ``prior_z_network_cls``
+                is not None, this network must be handle input with spec
+                ``(prior_input_tensor_spec, input_tensor_spec, prior_z_network.output_spec)``.
+            prior_input_tensor_spec: the input spec for ``prior_z_network``.
+            prior_z_network_cls: an encoding network that outputs a vector of logits
+                representing the a prior ``z`` distribution given the prior input.
+            mode: either 'st' or 'st-gumbel'.
+            gumbel_temp_scheduler: the temperature scheduler for gumbel-softmax.
+                Only used when ``mode=='st-gumbel'``.
+            beta: the weight for KL-divergence
+            target_kld_per_categorical: if not None, then this will be used as the
+                target KLD *per Categorical* to automatically tune beta.
+            beta_optimizer: if not None, will be used to train beta.
+            name (str):
+        """
+        Algorithm.__init__(self, name=name)
+
+        assert z_spec.is_discrete
+        self._n_categories = int(z_spec.maximum - z_spec.minimum + 1)
```

Comment:
Since the output z is always a vector and bounded in [0, n_gategories), to avoid misuse, we can simply assert z_spec.ndim==1 and z_spec.minimum==0.